Incorrect "unreachable code" warning

Using 2.13.10

```scala
object T {
  private sealed trait T
  private object O extends T
  private trait U extends T
  private object P extends U

  def t(t: T) = t match {
    case O => ()
    case _: U => println("hai")
  }

  def main(args: Array[String]): Unit = {
    t(P)
  }
}
```

gives

```
➜ sandbox sc Test.scala
Test.scala:9: warning: unreachable code
    case _: U => println("hai")
                        ^
1 warning

➜ sandbox s T
hai
```

The warning goes away when `U` is changed to `sealed`

--- a/src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala
+++ b/src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala
@@ -129,8 +129,11 @@ trait TreeAndTypeAnalysis extends Debugging {
       def filterAndSortChildren(children: Set[Symbol]) = {
         // symbols which are both sealed and abstract need not be covered themselves,
         // because all of their children must be and they cannot otherwise be created.
-        children.toList.filter(x => !(x.isSealed || x.isPrivate) || !x.isAbstractClass || isPrimitiveValueClass(x))
-          .sortBy(_.sealedSortName)
+        val children1 = children.toList.filterNot(child => child.isSealed && child.isAbstractClass).sortBy(_.sealedSortName)
+        children1.filterNot { child =>
+          // remove private abstract children that are superclasses of other children, for example in t6159 drop X2
+          child.isPrivate && child.isAbstractClass && children1.exists(sym => (sym ne child) && sym.isSubClass(child))
+        }
       }
 
       @tailrec def groupChildren(wl: List[Symbol], acc: List[List[Symbol]]): List[List[Symbol]] = wl match {

--- a/test/files/neg/t6159.check
+++ b/test/files/neg/t6159.check
@@ -1,0 +1,7 @@
+t6159.scala:10: warning: match may not be exhaustive.
+It would fail on the following input: (_ : A.this.X2)
+  def f(x: X) = x match {
+                ^
+error: No warnings can be incurred under -Werror.
+1 warning
+1 error

--- a/test/files/neg/t6159.scala
+++ b/test/files/neg/t6159.scala
@@ -1,0 +1,13 @@
+// scalac: -Werror
+// like test/files/pos/t6159.scala
+// but with T2 not private
+trait A {
+  sealed abstract class X
+  private class X1 extends X with X2 { }
+  trait X2 extends X
+  sealed trait X3 extends X
+
+  def f(x: X) = x match {
+    case _: X1 => 0
+  }
+}

--- a/test/files/pos/t12712.scala
+++ b/test/files/pos/t12712.scala
@@ -1,0 +1,16 @@
+// scalac: -Werror
+object T {
+  private sealed trait T
+  private object O extends T
+  private trait U extends T
+  private object P extends U
+
+  private def t(t: T) = t match {
+    case O => ()
+    case _: U => println("hai")
+  }
+
+  def main(args: Array[String]): Unit = {
+    t(P)
+  }
+}

